### ISSUE TITLE
Consider moving Unicode properties to separate library

### ISSUE THREAD

**Issue by facelessuser** [facelessuser/backrefs#135]
I'm not sure how useful it would be outside of Backrefs to most people, but I know we also use it in Wcmatch.

In Wcmatch we tap into uniprops to enable `[[:posix;]]` character classes, but we don't actually generate backref regex for 2 reasons:

1. We expose regular expression translation so the user can covert fnmatch/glob patterns directly to Python compatible regular expressions. We don't want to have to add a disclaimer "Oh, and you have to use Backrefs".

2. Wcmatch already has to parse the pattern once, and then feed it to Re which has to parse the pattern once. I'd hate to add yet another level of indirection of pattern parsing.

So, would it be helpful to push Unicode property handling to a separate library? And if so, what would that look like?

Currently we store all the properties in strings that are ready to be injected into regular expressions. We could expose a simple API that allows you to test characters etc.  Maybe that would be helpful?

```
uniprops.is_property('block=basic latin', 'a')
```

As far as backrefs is concerned, it is pretty much feature complete. If anything, I imagine features will be removed as Python's RE implements things like Unicode properties. Moving Uniprops out will cause less churn here probably.

The real question is whether  it is worth it. Wcmatch works fine as it is, but just wondering if it would be useful to anyone else and expose this stuff as a library. Something to think about.

**Comment by facelessuser** [comment]
To make this a practical library, we'd probably want to provide the look up tables in C. That is if we wanted it to be a general purpose library. Again, not sure how motivated I am to do that, but that is probably the way to go for a general purpose library that performs well. If we had functions that compared characters as it is now, we'd have to compile regular expressions to then compare the character...

In the end, considering that both applications in which we are using it is to generate regular expressions, maybe the way we have it now is fine. Wcmatch can keep taking advantage of the undocumented interface behind the scenes and backrefs continues to be what is now, a wrapper to augment regular expression patterns to expose features such as Unicode  and casing in replace patterns.

**Comment by facelessuser** [comment]
I'm reopening this as I think it is unfortunate that `wcmatch` has to include this whole package, but doesn't actually use anything but the Unicode table work. We need the Unicode table work in `wcmatch`, but we honestly don't need everything else in `backrefs`.

We could certainly expose a function that returns the `re` compatible expression instead of compiling it directly to regex and have `wcmatch` use that to justify the inclusion of `backrefs`, but we would just be adding an additional preprocessor step for the sole purpose of justifying the inclusion of the package. I'd rather split out the Unicode work and use it directly to keep `wcmatch` from taking a performance hit. And it makes no sense to expose the Unicode API directly in this package. Backrefs is simply an alternative regex wrapper.

I don't want to duplicate the Unicode work just so wcmatch can use it, but I think it is weird and confusing to people to see this package get included in `wcmatch`, and yet nothing from the public API is used.

If we split out the Unicode work into a separate repo, even if the only packages to use it are wcmatch and backrefs, that can provide value for us. It doesn't so much matter whether people at large find the package attractive as long as things like `wcmatch` and/or `backrefs` is found useful and it makes our lives easier.

To be honest, out of `backrefs` and `wcmatch`, I'd venture that `wcmatch` is the more popular package. `backrefs` is probably only downloaded as much as it is because `wcmatch` gets downloaded. This doesn't bother me as I still find this package useful even if no one else uses it.

---

We don't really have to try and make the Unicode package attractive to others, just a reusable package that plugs into `wcmatch` and `backrefs`. We can basically keep the API the same and simply state it is a regular expression Unicode package, or in other words, a package that can provide character groups formatted for regular expression character groups.

```py
re.compile(r'some pattern [' + uniprops.get_unicode_property('Letter') + ']')
```

It is not a concern to us whether people actually use this pacakge directly or not, frankly, this is what the uniprops code was built to do, and does it great for both `backrefs` and `wcmatch`. As long as it serves its intended purposes, I think it is worth it.

**Comment by facelessuser** [comment]
Cleaned up Unicode properties and have it split out and working locally. Adding tests for aliases exposed a number alias related bugs, so glad we got that in now.

Next is just getting the package up on pypi. I May end up tweaking the API a little to have to clean things up a little, but it should mostly remain the same.

**Comment by facelessuser** [comment]
#156 should close out this issue. We need to make sure `wcmatch` works with the new package and is separated from this one.

**Comment by facelessuser** [comment]
Looks like everything is working in this package and `wcmatch`, so the separation with refactoring worked out great. And we discovered and fixed a number of bugs along the way.

**Comment by facelessuser** [comment]
Before we relase a new version, we are going to dramatically improve testing in the `uniprops` library. At this point, in `backrefs`, the only bugs we run into are with Unicode properties, and that is mainly because we are not testing extensively enough. Generating the tables is very complicated, and we sometimes miss subtle things. The most complicated part of this project was/is the Unicode tables. I imagine with those moved out into a separate package that there will be even less churn in `backrefs`.

Part of the problem is that we don't actually commit the tables, so it is hard to see the differences. We don't commit them because we genrate different tables to match the version of supported Unicode in the varios versions of Python that we support. We can generally do better by Generating a folder with the old results and then make changes and compare against the old results. This is something I'll be doing when moving forward.

Second, we just need to ensure we inspect new tables a bit closer. This is harder to simply catch with tests. If you are generating the table from the Unicode text files, and your assumption is you've parsed them correctly, then your tests will be built with that understanding. The Vertical Orientation bug was an example of this, we should have visually inspected the table data points closer before releasing instead of just making sure it got populated.

Testing wise, we need to dot the following as these were unexpected bugs that shouldn't have happened:

- Test that all name and value combinations resolve through the API.
- Testing that all aliases resolve through the API.
- Test that all keys are cased properly in the structures.
- All structures in our Unicode tables are also present in the ASCII tables.

There is so much generated that it is hard to spot check everything, but these can easily be automated and ensure general integrity.

Bugs are part of programming, and parsing Unicode tables can be harder than it has to be sometimes due to how the data is presented at times, so I'm sure occasionally some issues will slip in, but I'd like to ensure moving forward that the dumbest bugs are stammpped out completely. I'd love to be able to focus on just updating the Unicode tables for new Unicode versions instead of fixing old generation 🙂.

**Comment by facelessuser** [comment]
So UniProps' testing is finished I believe and it has never been tested as thoroughly as it is now. TBH, it probably can't be tested too much more without reparsing the Unicode files just to make sure you are parsing them right the first time.

With that said, a couple more small cases arose where "unknown" aliases were backward and one interesting case in regards to order of evaluation.

The Regex library evaluates `gc`, `script`, `block`, and `binary` shortcuts in this order: `gc`, `script`, `block` and `binary`. I guess you could maybe argue that binary properties are not really shortcuts, they just have no values, so they seem like the shortcuts of `script` and/or `block` as you never actually specify them as `binary=name`, but I digress.

Perl on the other hand seems to evaluate `gc`, `binary`, `script`, and `block`. This actually makes more sense than what Regex documents that it does. You can actually have binary aliases that will never resolve, such as `idc` (`id_continue`) as there is a `block` property called `ideographic_description_characters that has an alias of `idc`.  If we look at Perl, they state:

```
  \p{Block: IDC}    \p{Block=Ideographic_Description_Characters} (NOT \p{ID_Continue} NOR \p{Is_IDC})
```

So, with that, we can see `binary` names should be ensured to resolve. Which makes sense as those are the names of the properties. If you have trouble with `block`, you can always use the more encouraged form of `blk: idc`. `Binary doesn't have alternate forms to express the same thing, so it should take priority over the `block` shortcut.

In short, I think the Python Regex library got this wrong, and we shouldn't follow down this path and instead should model Perl, which Regex is trying to provide a similar form of.

**Comment by facelessuser** [comment]
I've created an issue over at Regex here to at least see the author's opinion on this matter (order of evaluation of Unicode properties): https://bitbucket.org/mrabarnett/mrab-regex/issues/398/order-of-evaluation-of-unicode-properties

**Comment by facelessuser** [comment]
It is intersting and It was brought up that PCRE implementations in general may follow after Regex in relation to order of evaluation: GC, Script, Block, Binary. Maybe it does in fact make sense to continue forward with what PCRE does instead of Perl. I imagine they are less likely to break things like Perl does, but who knows 🤷.

**Comment by facelessuser** [comment]
And my earlier argument is a big flawed as you can specify binary properties as `\p{name}` or `\p{name: t}`, so you can remove the conflict regardless of the order.

**Comment by facelessuser** [comment]
So, the way it seems, Regex sounds like it follows closer to PCRE even though it borrows more from Perl than probably other regular expression engines. Most PCRE handles things like General Category and Script, and that is about it. Which makes sense why Regex favors Script right after GC. Block and Binary are just extra. It also explains why Regex favors Script instead of Script_Extensions when when using the shortcut form of script, as that is what PCRE does.

**Comment by facelessuser** [comment]
Turns out `wcmatch` never enabled Unicode support and has been using the ASCII definition in its character classes all this time. While the documents stated otherwise, we never enabled Unicode support and never tested for it. We only ever tested for ASCII and enabled ASCII.

No one ever complained, and I didn't notice in my use of it either. I don't think `wcmatch` is going to stick with ASCII tables and drop `backrefs` support. We don't need to split this out. We just need to ensure all the bugfixes are brought back over here.

**Comment by facelessuser** [comment]
I'm closing this issue as we are no longer going to move out the Unicode stuff. If I change the ordering of property evaluation, it will be in the next major release and not this upcoming bugfix release. The good news is that the Unicode stuff has never been so well tested, so I'm optimistic that we shouldn't have any more surprises.
